%% multiple_genome_alignment.tex
%% Author: Leighton Pritchard
%% Copyright: James Hutton Institute
%% Multiple genome alignment approaches

%
\begin{frame}
  \frametitle{Multiple genome alignments}
  \textcolor{olive}{Multiple genome alignments are ``harder'' than pairwise} \\
  \begin{itemize}
    \item Computationally difficult to produce
    \item \textcolor{red}{Lead to NP-complete optimisation problems!}
  \end{itemize}  
  \textcolor{hutton_green}{Solutions:} \textbf{heuristics}
  \begin{itemize}
    \item Progressive (build a tree, combine pairwise alignments)
    \item Iterative (realign initial sequences as new genomes added)
    \item \textcolor{hutton_blue}{Positional homology}
    \item \textcolor{hutton_purple}{\textit{Glocal} alignments}
  \end{itemize}  
\end{frame}

%
\begin{frame}
  \frametitle{Multiple genome alignment}
  Many tools either positional homology or glocal alignment
  \begin{columns}[T] 
    \column{.5\textwidth} 
      \textcolor{olive}{Several tools:}
      \begin{itemize}
        \item Mugsy: {\tiny\href{http://mugsy.sourceforge.net/}{(http://mugsy.sourceforge.net/)}}
        \item \textcolor{hutton_blue}{MLAGAN: {\tiny\href{http://lagan.stanford.edu/lagan_web/index.shtml}{(http://lagan.stanford.edu/lagan\_web/index.shtml)}}}
        \item TBA/MultiZ: {\tiny\href{http://www.bx.psu.edu/miller_lab/}{(http://www.bx.psu.edu/miller\_lab/)}}
        \item \textcolor{red}{\textbf{Mauve}: {\tiny\href{http://gel.ahabs.wisc.edu/mauve/}{(http://gel.ahabs.wisc.edu/mauve/)}}}
      \end{itemize}
    \column{.5\textwidth}
      \includegraphics[width=0.75\textwidth]{images/poshom_v_glocal}
  \end{columns}    
\end{frame}

%
\begin{frame}
  \frametitle{LAGAN
  \footnote{\tiny{\href{http://dx.doi.org/10.1101/gr.926603
}{Brudno \textit{et al.} (2003) \textit{Genome Res.} doi:10.1101/gr.926603
}}}
  }
  Rapid pairwise alignment of homologous genomes
  \begin{columns}[T] 
    \column{.55\textwidth} 
      \textcolor{olive}{Algorithm:}
      \begin{enumerate}
        \item \textcolor{hutton_green}{Generate local alignments {\small(\textit{anchors}, B)}}
        \item \textcolor{hutton_blue}{Construct rough global map {\small(\textit{max-scoring ordered subset}, C)}}
        \item Join anchors that lie within a threshold distance
        \item \textcolor{hutton_purple}{Compute global alignment within each anchor {\small(\textit{Needleman-Wunsch}, D)}}
      \end{enumerate}
    \column{.45\textwidth}
      \includegraphics[width=\textwidth]{images/lagan_method}
  \end{columns}    
\end{frame}

%
\begin{frame}
  \frametitle{MLAGAN
  \footnote{\tiny{\href{http://dx.doi.org/10.1101/gr.926603
}{Brudno \textit{et al.} (2003) \textit{Genome Res.} doi:10.1101/gr.926603
}}}
  }
  Multiple genome alignment of $k$ genomes in $k-1$ alignment steps \\
  Progressive/iterative, using phylogenetic tree (like \texttt{CLUSTAL})
  \begin{columns}[T] 
    \column{.55\textwidth} 
      \textcolor{olive}{Algorithm:}
      \begin{enumerate}
        \item \textcolor{hutton_green}{Construct rough global maps between each pair of sequences {\small(step C in LAGAN)}}
        \item \textcolor{hutton_blue}{Progressive multiple alignment with anchors:}
		\begin{itemize}
          \item \textcolor{hutton_purple}{LAGAN alignment between closest pair of sequences: combined as \textit{multi-sequences}}
          \item \textcolor{hutton_purple}{Find rough global maps of this multi-sequence to all other multi-sequences}          
        \end{itemize}
      \end{enumerate}
    \column{.45\textwidth}
      \includegraphics[width=\textwidth]{images/mlagan_method}
  \end{columns}    
\end{frame}

%
\begin{frame}
  \frametitle{Human:Mouse:Rat}
  \begin{center}
    \includegraphics[height=0.7\textheight]{images/mouse-rat-photo}
  \end{center}
\end{frame}

%
\begin{frame}
  \frametitle{Human:Mouse:Rat
  \footnote{\tiny{\href{http://dx.doi.org/10.1101/gr.2067704
}{Brudno \textit{et al.} (2004) \textit{Genome Res.} doi:10.1101/gr.2067704
}}}
  }
  \textcolor{hutton_green}{Three-way progressive alignment} 
  \begin{columns}[T] 
    \column{.4\textwidth} 
    \textcolor{hutton_blue}{Initial alignments: BLAT \\
    Synteny: LAGAN/MLAGAN \\~\\}
    \textcolor{hutton_purple}{Three-way synteny}
      \begin{itemize}
        \item Homologous (HMR)
        \item Rodent-only (MR)
        \item Human-mouse (HM)
        \item Human-rat (HR)
      \end{itemize}
    \column{.6\textwidth}
      \begin{center}
        \includegraphics[width=\textwidth]{images/hmr_alignment} \\
        {\tiny(synteny mapped to rat genome)}
      \end{center}
  \end{columns}    
\end{frame}

%
\begin{frame}
  \frametitle{MAUVE
  \footnote{\tiny{\href{http://dx.doi.org/10.1101/gr.2289704
}{Darling \textit{et al.} (2003) \textit{Genome Res.} doi:10.1101/gr.2289704
}}}
  }
  MAUVE/Progressive MAUVE: \href{http://gel.ahabs.wisc.edu/mauve/}{http://gel.ahabs.wisc.edu/mauve/} \\
  \begin{columns}[T] 
    \column{.55\textwidth} 
      \textcolor{RawSienna}{Algorithm:}
      \begin{enumerate}
        \item \textcolor{hutton_green}{Find local alignments (\textit{multi-MUMs}, A)}
        \item Build guide tree from multi-MUMs
        \item \textcolor{hutton_blue}{Select subset of multi-MUMs (\textit{anchors}, B)}
        \begin{itemize}
          \item Partition into \textit{local collinear blocks} (LCBs)
        \end{itemize}
        \item \textcolor{hutton_purple}{Recursive anchoring to refine anchors (C)}
        \item Progressive alignment against guide tree
      \end{enumerate}
    \column{.45\textwidth}
      \includegraphics[width=0.9\textwidth]{images/mauve_method}
  \end{columns}    
\end{frame}

%
\begin{frame}
  \frametitle{MAUVE
  \footnote{\tiny{\href{http://dx.doi.org/10.1101/gr.2289704
}{Darling \textit{et al.} (2003) \textit{Genome Res.} doi:10.1101/gr.2289704
}}}
  }
  MAUVE alignment of LCBs in nine enterobacterial genomes \\
  \textcolor{hutton_green}{Evidence for rearrangement of homologous backbone sequence}
  \begin{center}
    \includegraphics[height=0.6\textheight]{images/mauve_entero}
  \end{center}    
\end{frame}

%
\begin{frame}
  \frametitle{Draft genome alignment}
  \textcolor{olive}{High-throughput genome assemblies may be fragments (contigs)} \\
  Contigs can be ordered (\textit{scaffolded}):
  \begin{itemize}
    \item \textcolor{hutton_blue}{without alignment}, by long or paired-end reads
    \item \textcolor{hutton_purple}{by alignment}, to complete \textit{reference} genomes
    \item \textcolor{hutton_purple}{by alignment}, to other draft incomplete genomes
  \end{itemize}
  \begin{center}
    \includegraphics[width=\textwidth]{images/mauve_scaffolding}
  \end{center}    
\end{frame}

%
\begin{frame}
  \frametitle{Draft genome alignment}
  Alignment-based scaffolding:
  \begin{itemize}
    \item \textcolor{hutton_blue}{MUMmer}: \texttt{nucmer}, or \texttt{promer} (if divergent)
    \item \textcolor{hutton_purple}{MAUVE/Progressive MAUVE}: Mauve Contig Mover (MCM)
  \end{itemize}
  \begin{center}
    \includegraphics[width=\textwidth]{images/mauve_scaffolding}
  \end{center}    
\end{frame}

%
\begin{frame}
  \frametitle{Multiple genome alignment}
  \Large{
    \textcolor{hutton_blue}{
      \textbf{
      EXERCISE 5: \\
      {\small \href{https://github.com/widdowquinn/Teaching-2015-03-17-UoD_compgenvis/blob/master/exercises/whole_genome_alignment/whole_genome_alignments_B.md}{\texttt{whole\_genome\_alignment/whole\_genome\_alignments\_B.md}}}
      }
    }
  }
\end{frame}

%
\begin{frame}
  \frametitle{Multiple genome alignment}
  \Large{
    \textcolor{hutton_blue}{
      \textbf{
      EXERCISE 6: \\
      \texttt{ex06\_biopython\_visualisation.ipynb}
      }
    }
  }
\end{frame}

% DNA-DNA hybridisation
\begin{frame}
  \frametitle{Chromosome painting\footnote{\tiny{\href{http://dx.doi.org/10.1093/molbev/mst055}{Yahara \textit{et al}. (2013) \textit{Mol. Biol. Evol.} \textbf{30}:1454-1464 doi:10.1093/molbev/mst055}}}}
  ``Chromosome painting'' infers recombination-derived `chunks'\\
  Genome's haplotype constructed in terms of recombination events from a `donor' to a `recipient' genome\\
  \begin{center}
    \includegraphics[width=0.7\textwidth]{images/chromosome_painting}
  \end{center}     
\end{frame}

% DNA-DNA hybridisation
\begin{frame}
  \frametitle{Chromosome painting\footnote{\tiny{\href{http://dx.doi.org/10.1093/molbev/mst055}{Yahara \textit{et al}. (2013) \textit{Mol. Biol. Evol.} \textbf{30}:1454-1464 doi:10.1093/molbev/mst055}}}}
  Recombination events summarised in a \textit{coancestry matrix}.\\
  \textit{H. pylori}: most within geographical bounds, but asymmetrical donation from Amerind/East Asian to European isolates.
  \begin{center}
    \includegraphics[width=0.75\textwidth]{images/coancestry}
  \end{center}     
\end{frame}